Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp

```diff
@@ -244,41 +404,78 @@ CDMInstanceOpenCDM::Session::Session(const media::OpenCdm& source, Ref<WebCore::
 
     m_message = SharedBuffer::create(message.c_str() + offset, message.size() - offset);
     m_needsIndividualization = requestType.length() == 1 && static_cast<WebCore::MediaKeyMessageType>(requestType.toInt()) == CDMInstance::MessageType::IndividualizationRequest;
+    for (const auto& challengeCallback : m_challengeCallbacks)
+        challengeCallback(this);
+    m_challengeCallbacks.clear();
+}
+
+void CDMInstanceOpenCDM::Session::didUpdateKey(const std::string& status)
+{
+    for (auto& keyStatusData : m_keyStatusDatas) {
+        m_lastStatus = opencdm_session_status(m_session.get(), keyStatusData.m_key.get(), keyStatusData.m_keySize);
+        keyStatusData.m_callback(this, m_lastStatus, String(status.c_str(), status.empty() ? 0 : status.size()));
+    }
+    m_keyStatusDatas.clear();
+
+    for (auto& loadCallback : m_loadCallbacks) {
+        loadCallback(this, true, String(status.c_str(), status.empty() ? 0 : status.size()));
+    }
+    m_loadCallbacks.clear();
+
+    for (auto& removeCallback : m_removeCallbacks) {
+        removeCallback(this, true, String(status.c_str(), status.empty() ? 0 : status.size()));
+    }
+    m_removeCallbacks.clear();
+}
+
+void CDMInstanceOpenCDM::Session::didReceiveMessage(const std::string& message)
+{
+}
+
+void CDMInstanceOpenCDM::Session::generateChallenge(DidGenerateChallenge&& callback)
+{
+    if (isValid()) {
+        callback(this);
+        return;
+    }
+
+    m_challengeCallbacks.append(WTFMove(callback));
 }
 
-std::pair<media::OpenCdm::KeyStatus, String> CDMInstanceOpenCDM::Session::update(const uint8_t* data, const unsigned length)
+void CDMInstanceOpenCDM::Session::update(const uint8_t* data, const unsigned length, DidUpdateKey&& callback)
 {
-    std::string response;
-    m_lastStatus = m_session.Update(data, length, response);
-    return std::make_pair(m_lastStatus, String(response.c_str(), response.empty() ? 0 : response.size()));
+    KeyUpdateData updateData(WTFMove(callback), data, length);
+    m_keyStatusDatas.append(WTFMove(updateData));
+    if (opencdm_session_update(m_session.get(), data, length)) {
+        didUpdateKey(std::string());
+    }
 }
 
-std::pair<bool, String> CDMInstanceOpenCDM::Session::load()
+void CDMInstanceOpenCDM::Session::load(DidChangeData&& callback)
 {
-    std::string response;
-    bool result = !m_session.Load(response);
-    return std::make_pair(result, String(response.c_str(), response.empty() ? 0 : response.size()));
+    m_loadCallbacks.append(WTFMove(callback));
+    if (!opencdm_session_load(m_session.get()))
+         didUpdateKey(std::string());
```

Comment:
Why have we updated keys in a load failure?

Reply:
didUpdateKey() simply does all the work and it should normally be invoked in case all goes ok but since load fails I call it directly.

Reply:
Is that ok or I should add some wrapper making it more readable? I think loadError() calling didUpdateKey()  (same would apply to remove())

Reply:
Yeah, if it's low effort to make that a little clearer I would go for it. We'll see what @calvaris thinks when he's got time to review. I'd be happy with this landing with or without the change.